Assets/MixedRealityToolkit.Tests/PlayModeTests/InputEventSystemTests.cs

```diff
@@ -0,0 +1,300 @@
+{
+    class InputEventSystemTests
+    {
+        // Can't use NUnit setup routine, because needs to skip several frames while initializing/removing objects.
+        public IEnumerator SetupMrtkWithoutCursors()
+        {
+            TestUtilities.InitializeMixedRealityToolkitAndCreateScenes(true);
+            TestUtilities.InitializePlayspace();
+
+            IMixedRealityInputSystem inputSystem = null;
+            MixedRealityServiceRegistry.TryGetService(out inputSystem);
+
+            Assert.IsNotNull(inputSystem, "Input system must be initialized");
+
+            // Let input system to register all cursors and managers.
+            yield return null;
+
+            // Switch off / Destroy all input components, which listen to global events
+            Object.Destroy(inputSystem.GazeProvider.GazeCursor as Behaviour);
+            inputSystem.GazeProvider.Enabled = false;
+
+            var diagnosticsVoiceControls = Object.FindObjectsOfType<DiagnosticsSystemVoiceControls>();
+            foreach (var diagnosticsComponent in diagnosticsVoiceControls)
+            {
+                diagnosticsComponent.enabled = false;
+            }
+
+            // Let objects be destroyed
+            yield return null;
+
+            // Check that input system is clean
+            Assert.AreEqual(((BaseEventSystem)inputSystem).EventHandlersByType.Count, 0, "Input event system handler registry is not empty in the beginning of the test.");
+
+            yield return null;
+        }
+
+        [TearDown]
+        public void ShutdownMrtk()
+        {
+            TestUtilities.ShutdownMixedRealityToolkit();
+        }
+
+        /// <summary>
+        /// </summary>
+        [UnityTest]
+        public IEnumerator Test01_TestObjectBasedEventRegistration()
+        {
+            yield return SetupMrtkWithoutCursors();
+
+            IMixedRealityInputSystem iInputSystem = null;
+            MixedRealityServiceRegistry.TryGetService(out iInputSystem);
+
+            BaseEventSystem inputSystem = (BaseEventSystem)iInputSystem;
+
+            var object1 = new GameObject("Object");
+
+            // Second handler weirdly depends on the first one due to event registration working on the entire object
+            var objectBasedListener = object1.AddComponent<TestInputGlobalListenerObjectBased>();
+            var handlerBasedListener = object1.AddComponent<TestInputGlobalListenerHandlerBasedSpeechHandler>();
+
+            yield return null;
+
+            // Event listener collection is filled for backward compatibility
+            Assert.AreEqual(inputSystem.EventListeners.Count, 1, "Event listener for old event system API hasn't been registered");
+            Assert.Contains(objectBasedListener.gameObject, inputSystem.EventListeners, "Wrong event listener registered for old event system API.");
+
+            Assert.AreEqual(inputSystem.EventHandlersByType.Count, 2, "Input event system doesn't contain expected event handler types.");
```

Comment:
If there's any way for us to use an autoformatter to reign in the line length here it might be helpful (reviewing using github actually is difficult because it's longer that CR viewport width)

Reply:
I think it should be much better now with Collection comparisons.